feat: integrate the NodeIdProvider in the application startup

entangled90 opened the pull request:
## Description
In this PR the `NodeIdProvider` is integrated in the rest of the application in the `dist` and `configuration` modules.
A new configuration is introduced to configure the `S3NodeIdRepository` in such a way that it can be extended later on if different providers need to be added (e.g. RDBMS, azure, gcp).
The configuration is backward compatible:
- `camunda.cluster.node-id` is now nullable, to allow for better validation of the configurations.
- `camunda.cluster.dynamic-node-id` is introduced with a `.s3` section for `S3NodeIdRepository`.
- `camunda.cluster.dynamic-node-id.type` is used to discern between the possible implementation, defaulting to `none`.
  if set to `none`, then the configuration is not used and we default to the current behaviour, with `node-id=0` if not set.

An IT test is added that verifies that a cluster correctly starts up using `S3NodeIdRepository` with `LocalStack` as S3 implementation.




## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

relates #40476

entangled90 committed:
refactor: Create NodeIdProvider interface

With an interface, we can create a "static" implementation to be used
when the nodeId is set in the configuration

entangled90 committed:
feat: integrate NodeIdProvider in the application with a static value

entangled90 committed:
feat: nodeId in cluster configuration is nullable

entangled90 committed:
refactor: rename expiryDuration to leaseDuration in S3NodeRepository.Config

entangled90 committed:
feat: add section in Cluster unified configuration for dynamic node id

entangled90 commented on dist/src/main/java/io/camunda/zeebe/broker/NodeIdProviderConfiguration.java:
```diff
@@ -0,0 +1,38 @@
+  }
+
+  @Bean
+  public NodeIdProvider staticNodeIdProvider() {
```
Yes, I tried to just set a static one to check I didn't break anything. In the next commits, I'll integrate the dynamic version based on config.

entangled90 replied:
Considering we already use the unified config, we only have one property, is it ok to use `@ConditionalOnProperty` or do we have other tools available?

entangled90 committed:
feat: when nodeId and dynamicNodeId are not configured, default to set nodeId=0

entangled90 committed:
refactor: NodeIdRepository does not initialize in the constructor

Initializing the repository in a separate method enable us to remove the
field `clusterSize` in the class, as it's only needed when initializing.
This is safer, considering that the clusterSize in the configuration may
not reflect the actual size of the cluster.

The repository is initialized by the NodeIdProvider directly as well.

entangled90 committed:
fix: extra null checks in RepositoryNodeIdProvider constructor

entangled90 committed:
fix: if the bucket does not exist, it should fail fast

entangled90 committed:
feat: add taskId in camunda.cluster.dynamicNodeId.s3 config

entangled90 committed:
test: add a QA test for a single node acquiring the lease in S3

entangled90 committed:
test: migrate single application cluster to TestCluster

entangled90 committed:
refactor: optional arguments are wrapped in Optional

entangled90 committed:
build: fix missing dependencies

entangled90 committed:
test: remove unnecessary properties for webapps

entangled90 committed:
docs: fix JavaDoc

entangled90 committed:
fix: missing whitespace in error message

deepthidevaki commented on zeebe/dynamic-node-id-provider/src/main/java/io/camunda/zeebe/dynamic/nodeid/RepositoryNodeIdProvider.java:
```diff
@@ -0,0 +1,206 @@
+  @Override
+  public void initialize(final int clusterSize) {
+    nodeIdRepository.initialize(clusterSize);
+    CompletableFuture.runAsync(() -> acquireInitialLease(clusterSize), executor)
```
:question: This is not waiting until the lease is acquired. How is it ensured that, the broker do not start before the nodeId is assigned?

deepthidevaki commented on qa/acceptance-tests/src/test/java/io/camunda/it/cluster/DynamicNodeIdTest.java:
```diff
@@ -0,0 +1,142 @@
+            .collect(Collectors.toSet());
+    assertThat(configuredTaskIds).allSatisfy(o -> assertThat(o).isNotNull());
+    assertThat(leases.stream().map(Lease::taskId).collect(Collectors.toSet()))
+        .isEqualTo(configuredTaskIds);
```
:x: Would it be possible to verify that the broker id is assigned correctly from the leased node id?

deepthidevaki replied:
We should also verify the cluster is formed correctly and check the topology is complete.

deepthidevaki commented on qa/acceptance-tests/src/test/java/io/camunda/it/cluster/DynamicNodeIdTest.java:
```diff
@@ -0,0 +1,142 @@
+
+  @AutoClose private static S3Client s3Client;
+  private static String bucketName;
+  private static String taskId;
```
:wrench: May be rename it to `taskIdPrefix`? I thought you were using the same task id for all nodes for a moment.

deepthidevaki replied:
I mean the variable name here in the test. Not the actual configuration property. This is used as a prefix to generate the taskId per broker.

deepthidevaki commented on configuration/src/main/java/io/camunda/configuration/Cluster.java:
```diff
@@ -78,7 +80,10 @@ public class Cluster implements Cloneable {
+  private Integer nodeId;
+
+  @NestedConfigurationProperty
+  private DynamicNodeIdConfig dynamicNodeId = new DynamicNodeIdConfig();
```
:wrench: May be add a documentation when and how to use this configuration.

deepthidevaki reviewed (changes requested):
Thanks. The integration via spring beans injections looks good. 

- :x: The nodeId in the broker cfg is not overwritten with the acquired nodeId. The IT test does not catch this. 
- :wrench: Could you add a bit more context to the documentation of the configuration properties? eg:- why and when to set the properties etc.?

entangled90 committed:
fix: wait for initialization of NodeIdProvider

entangled90 committed:
fix: set nodeId in BrokerCfg.ClusterCfg based on NodeIdProvider value

Additional assertions in DynamicNodeIdTest verify that the nodeId <->
taskId mapping in S3 and in the broker configuration match.
The added assertions were not met before fix.

entangled90 committed:
docs: Better JavaDoc

entangled90 committed:
test: make some fields static in DynamicNodeIdTest

npepinpe commented on qa/acceptance-tests/src/test/java/io/camunda/it/cluster/DynamicNodeIdTest.java:
```diff
@@ -0,0 +1,162 @@
+  private static final int CLUSTER_SIZE = 3;
+
+  @TestZeebe
+  protected TestCluster testCluster =
```
❓ Will this be sub-classed?

npepinpe commented on qa/acceptance-tests/src/test/java/io/camunda/it/cluster/DynamicNodeIdTest.java:
```diff
@@ -0,0 +1,162 @@
+        testCluster.brokers().values().stream()
+            .map(
+                app ->
+                    app.property("camunda.cluster.dynamic-node-id.s3.taskId", String.class, null))
```
❓ Isn't it easier to set explicit task IDs in the test set up than dealing with nulls? 🤷

entangled90 replied:
This is for getting the value of the property, not for setting it.

npepinpe replied:
No I get it, I meant the assertions seemed to test that we also have one set there, because then we were asserting that there was something that we got :smile:

npepinpe commented on qa/acceptance-tests/src/test/java/io/camunda/it/cluster/DynamicNodeIdTest.java:
```diff
@@ -0,0 +1,162 @@
+                app ->
+                    app.property("camunda.cluster.dynamic-node-id.s3.taskId", String.class, null))
+            .collect(Collectors.toSet());
+    assertThat(configuredTaskIds).allSatisfy(o -> assertThat(o).isNotNull());
```
🔧 AssertJ has methods for all these things. In your case here, you have `doesNotContainNull`, for example.

npepinpe commented on zeebe/dynamic-node-id-provider/src/main/java/io/camunda/zeebe/dynamic/nodeid/RepositoryNodeIdProvider.java:
```diff
@@ -0,0 +1,207 @@
+            LOG.debug(
+                "Attempt to acquire the lease failed for all nodeIds, sleeping {} and retrying again",
+                currentDelay);
+            Thread.sleep(currentDelay);
```
❌ Whoa wait, very suspicious 😄 Must be use a thread sleep here?

entangled90 replied:
It's already a task running inside the executor. Since this is part of the initialization process (which is synchronous anyway) we just block directly here.
Alternatively, you could use retries resubmitting to the executor, but in the end you would still block in the spring initialization thread anyway, so :shrug:

npepinpe commented on zeebe/dynamic-node-id-provider/src/main/java/io/camunda/zeebe/dynamic/nodeid/RepositoryNodeIdProvider.java:
```diff
@@ -0,0 +1,207 @@
+        // release is already submitted to the executor, we can shut it down gracefully.
+        executor.shutdown();
+        // shutdown is taking too much time, let's shut it down by interrupting running tasks.
+        if (!executor.awaitTermination(5, TimeUnit.SECONDS)) {
```
❓ Could this make shutdown slow in some tests?

entangled90 replied:
In 99.9% of the tests, this NodeIdProvider is not used. The time taken to shutdown the executor is if a running task is blocking. 
We use this executor basically just to make S3 requests to `LocalStack` so it should be ~ 100-200 millis in most cases.

entangled90 replied:
Mostly because we are spawning just 1 thread and with jdk21 [JEP-491](https://openjdk.org/jeps/491) is not fixed yet. If you say we can ignore it, I can promote it to a virtual thread executor.

npepinpe reviewed (changes requested):
🚀 

- 🔧 Generally not a fan of having something like `nodeId` and then `DynamicNodeConfig`. Conceptually, static is just a way to configure that never changes, so I'd merge them into a single node ID configuration, and replace `Type.NONE` for `Type.STATIC` or something.
- Lots of comments, but this is a rather big PR. I think it could've been split personally, just OTOH: introduce config, then make NodeIdProvider an API (and adapt the rest), then add the repository provider and the integration tests 🤷 

Also I won't always review, but I'm very curious about this here. So apologies for the drive-by review, if this is annoying I can also scale back and just pester people on Slack 😬

entangled90 commented on qa/acceptance-tests/src/test/java/io/camunda/it/cluster/DynamicNodeIdTest.java:
```diff
@@ -0,0 +1,162 @@
+        .hasSize(CLUSTER_SIZE);
+
+    assertThat(s3NodeIdMapping).isEqualTo(configNodeIdMapping);
+  }
```
Good point, it may even be simpler to assert with this:
```java

   final var expectedMappings =
        IntStream.range(0, CLUSTER_SIZE)
            .boxed()
            .collect(Collectors.toMap(Function.identity(), DynamicNodeIdTest::taskId));
    assertThat(s3NodeIdMapping).isEqualTo(configNodeIdMapping).isEqualTo(expectedMappings);
```

entangled90 replied:
No it's not equivalent. 
```java
    assertThat(s3NodeIdMapping)
        .hasSize(CLUSTER_SIZE)
        .containsKeys(0, 1, 2)
        .containsValues("nodeId=0", "nodeId=1", "nodeId=2")
```
This assertion is generally false. I see how the taskId names are a bit misleading and it's probably better to just generate them as random `UUID`

entangled90 commented on zeebe/dynamic-node-id-provider/src/main/java/io/camunda/zeebe/dynamic/nodeid/RepositoryNodeIdProvider.java:
```diff
@@ -0,0 +1,207 @@
+  private final String taskId;
+  private final Runnable onLeaseFailure;
+  private final ScheduledExecutorService executor;
+  private final ExponentialBackoff backoff;
```
This was already merged in `main`, I will fix it once this PR is merged as it's a bit hard to keep track of differences.

entangled90 commented on zeebe/dynamic-node-id-provider/src/main/java/io/camunda/zeebe/dynamic/nodeid/RepositoryNodeIdProvider.java:
```diff
@@ -0,0 +1,207 @@
+      currentLease = tryAcquire(storedLease);
+    }
+    if (currentLease != null) {
+      LOG.info(
```
This is the first lease acquired, so it contains the nodeId, version & taskId. I think that since it's just logged once it contains vital information if we need to analyze what happens. Note that concurrency bugs might not repeat that often, so it will be helpful to have this information instead of redeploying at DEBUG and then wait for another occurence.

entangled90 committed:
test: implement suggestions for DynamicNodeIdTest

npepinpe reviewed (approved):
Approving my comments, let's wait for Deepthi to approve too :)

entangled90 committed:
feat: log line at DEBUG instead of INFO

entangled90 committed:
test: add ability to not set nodeId in TestClusterBuilder

entangled90 committed:
feat: when nodeId from old config is null, call nodeIdProvider

This is a temporary fix until we can integrate UnifiedConfiguration
inside the TestStandaloneBroker/TestCluster

entangled90 committed:
refactor: NodeIdProvider.initialize() returns a CompletableFuture<Void>

deepthidevaki reviewed (approved):
Thanks. Please check my comment before merging. I'm pretty sure we have to wait for the initialization to complete. 

Could you please create an issue to revert the changes as well?

entangled90 commented:
Create [40827](https://github.com/camunda/camunda/issues/40827)
